influxdb/src/main/scala/akka/stream/alpakka/influxdb/impl/InfluxDBFlowStage.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.influxdb.impl
+
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import akka.stream.alpakka.influxdb.{InfluxDBWriteMessage, InfluxDBWriteResult}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import org.influxdb.InfluxDB
+import org.influxdb.impl.InfluxDBMapper
+
+import scala.collection.immutable
+import org.influxdb.BatchOptions
+import org.influxdb.dto.Point
+
+/**
+ * INTERNAL API
+ */
+private[influxdb] class InfluxDBFlowStage[T, C](
+    clazz: Option[Class[T]],
+    influxDB: InfluxDB
+) extends GraphStage[FlowShape[immutable.Seq[InfluxDBWriteMessage[T, C]], immutable.Seq[InfluxDBWriteResult[T, C]]]] {
+  private val in = Inlet[immutable.Seq[InfluxDBWriteMessage[T, C]]]("in")
+  private val out = Outlet[immutable.Seq[InfluxDBWriteResult[T, C]]]("out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    clazz match {
+      case Some(c) => new InfluxDBMapperRecordLogic
+      case None => new InfluxDBRecordLogic
+    }
+
+  sealed abstract class InfluxDBLogic extends GraphStageLogic(shape) with InHandler with OutHandler {
+
+    protected def write(messages: immutable.Seq[InfluxDBWriteMessage[T, C]]): Unit
+
+    setHandlers(in, out, this)
+
+    override def onPull(): Unit = if (!isClosed(in) && !hasBeenPulled(in)) pull(in)
+
+    override def onPush(): Unit = {
+      val messages = grab(in)
+      if (messages.nonEmpty) {
+
+        influxDB.enableBatch(BatchOptions.DEFAULTS)
+        write(messages)
+        val writtenMessages = messages.map(m => new InfluxDBWriteResult(m, Option.empty))
+        influxDB.close()
+        push(out, writtenMessages)
+      }
+
+      tryPull(in)
+    }
+  }
+
+  final class InfluxDBRecordLogic extends InfluxDBLogic {
+
+    override protected def write(messages: immutable.Seq[InfluxDBWriteMessage[T, C]]): Unit =
+      messages.foreach {
+        case InfluxDBWriteMessage(point: Point, _) => {
+          influxDB.write(point)
```

Comment:
I'm a little bit confused about how we are using this [`influxdb-java`](https://github.com/influxdata/influxdb-java) library. I apologize if these questions are basic, but I am not familiar with this library.

When we call `influxDB.write(point)`, what is that actually doing? Is it just adding the message to an internal buffer? Does it perform a synchronous write?

How does it handle write failures? Are we guaranteed that the previous write succeeded when we emit this element?

Reply:
When we call to write a point and batch is enabled, then the write calls shall be batched until close is called.
Therefore in this case the messages are being written to the internal buffer.
Once close is called, then the messages are being flashed.

Reply:
So we need to call `close` in order to actually make the HTTP request to Influx? And when `close` returns, do we know if Influx has received the request (did it return an HTTP 201)?

Reply:
Exactly, once close is called the actual request is being made.
If there is an error during the request the driver throws an Exception.
The exception might be an InfluxDBException which has to do with the state or an InfluxDBIOException which has to do with the IO.

Reply:
Edit: my previous comment is misleading. Apart from the close operation which flushes the messages, there is also an interval that writes the messages to the database.

Reply:
I am still a bit confused about what happens when we call `close`. It appears to me that one of the two is occurring:

1. The thread will *block* while it makes the HTTP call. If Influx accepts the write, the `close` function will return, therefore we can emit the messages we have written. If the write fails, an exception will be thrown, and we can handle that in the decider.
2. The thread does not block, and the write occurs asynchronously. 

If 1) is correct, we need to make sure we are handling the thread blocking appropriately. This is outside my knowledge range, maybe @ennru can help you more around this. But we don't want one thread with a blocking call to cause thread starvation on other threads. If we need to use blocking, we should run this on a separate dispatcher that is optimized for blocking calls. I understand the Influx Library has its own thread pool solution, but we still call `close` on our thread pool, so we have to be careful.

If 2) is correct, we don't have "at least once delivery", because we emit the messages before we are sure that Influx has acknowledged them. 

Can you help clarify which of these is occurring?

Reply:
1) is correct.

I will explain the flow and why option 1 is correct.
Batching is enabled (influxDB.enableBatch()). 
By enabling batching all the points submitted for a write will be added to a queue.
By enabling batching a scheduler is configured, to run (by default every 1 second) and offload some of the messages in the queue to InfluxDB. It is clear that this step occurs in another thread managed by the driver.
Once batching is stop ( influxDB.close()) the scheduler is stopped and the points left on the queue are going to be written to the database (all of the batched in one request). This step is happening by the same thread that calls the close operation. Therefore the batch of points are written by the main thread.

Why '1)' is correct, and how the writes occur in the main thread?
In the connector code once batch is enabled the only action that happens is to add the messages to the queue and once they are added the batch operation is closed. Thus the point are send to persist through the main thread. The write and close steps occur before the scheduler is called at all which leads to the actions to take place on the main thread.

However considering the complexity of the above, which might lead to misconceptions and false usage, I shall make a commit which implements directly the actual functionality needed to sent all of the writes once off without using the drivers batch functionality (therefore without the need to deal and be aware of the internal mechanisms of the driver).

Reply:
If any of the operations called are known to be blocking, we'd want to make the stage logic run on the IO dispatcher.

Reply:
I updated the code and removed any parts that have to do with the drivers batching operation based on external threads . The blocking operations are isolated and thus the logic shall be set to run on the IO dispatcher.